### ISSUE TITLE
neverball and neverputt visuall diffs

### ISSUE THREAD

**Issue by kas1e** [ptitSeb/gl4es#57]
Version in testing 1.5.4, on amigaos4.

So noticed 2 issues.

Fist issues in neverball when we first time running game. After we run it , for a while (3 seconds on my setup) we have only menu and black background instead of background animation. Then, after 3 seconds background strangely blinks, and animation appear to be visibly. That only happens on running. When we go to options and then back, then background's animation always start as it, and all looks fine.

I do test on MiniGL, and there after running i can see how animation background appear together with menu , with a little fade in effect. So probably some state or something cause that.

Second issues, is in neverputt. To reproduce it just run it , choice "play", choice "firt" level (that one at left top side), choice 1 pplayer and run it. Now, instead of "ball", you see only shadow of ball. 

I do check it over minigl as well too, and ball is here. 

Can make screenshots for both issues if there is needs for.

**Comment by kas1e** [comment]
Basically with neverputt you can choice any level, there always no ball, and only shadow (at least i tried 3 different levels).

**Comment by kas1e** [comment]
And found another issue in neverball : in the starting anymation, there is no "coins" visibly. Also in all the levels coins have no texture , and they just empty.

**Comment by ptitSeb** [comment]
I do have coins on the Pandora, for sure (and it may be linked to the invisible ball in neverputt). So maybe the texture as some bigendian issue? Not sure what format they are.

Some traces in `src/gl/texture.c` might help.

**Comment by kas1e** [comment]
neverball:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_texture.txt

neverputt:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_texture.txt

In both games i just go to the first level, and exit from.

**Comment by ptitSeb** [comment]
Nope, it's not that. All texture are simple GL_RGB(A)/GL_UNSIGNED_BYTE, no issues here.

On Neverball, do you have the reflexion on the coins?

**Comment by kas1e** [comment]
Making some screenshots which probably better show what i mean. 

That how looks like menu in neverball (see there is no coins on background animation):
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_gl4es_menu.jpg

And that how it should looks like and how it looks like on minigl (see there is coin):
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_mgl_menu.jpg

Now, that how it looks like in game (see coins empty, with no textures, they white/transparently):
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_gl4es_ingame.jpg

And how it should be:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_mgl_ingame.jpg

And, in neverputt there is no "ball", like this:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_gl4es_ingame.jpg

While should be like this:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_mgl_ingame.jpg

Did i understand right, that for you, on pandora, everything as i show on minigl screenshots, and not like i have in my gl4es versions ?

**Comment by ptitSeb** [comment]
Yes, everything is fine on the Pandora: coins are there.
I can see on the screenshot in gameplay, that the coins seems to be there, has "ghosts". Because the texture of the coin is white, I assume there are some color stuff (to have yellow, red and blue coins), maybe there is some bigendian stuff on color this time...
I have done a GLES2 capture on the Pandora, and with neverball source, I'll try to understand how thoses coins are drawn.

**Comment by ptitSeb** [comment]
Sorry, but I see nothing suspicious on how thoses coins are drawn. It using Lighting and Material, with a single texture. I don't explain why it would be transparent.

Some example of shader used:
Vertex
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewMatrix;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute lowp vec4 _gl4es_Color;
attribute highp vec4 _gl4es_MultiTexCoord0;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelParameters {
  vec4 ambient;
};
uniform _gl4es_LightModelParameters _gl4es_LightModel;
struct _gl4es_MaterialParameters
{
   vec4 emission;
   vec4 ambient;
   vec4 diffuse;
   vec4 specular;
   float shininess;
};
uniform _gl4es_MaterialParameters _gl4es_FrontMaterial;
uniform _gl4es_MaterialParameters _gl4es_BackMaterial;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=0, separate=1, color_material=1)
// secondary=0, planes=000001
// texture=00000001, point=0
varying vec4 Color;
uniform highp vec4 _gl4es_ClipPlane_0;
varying mediump float clippedvertex_0;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;
varying vec4 SecColor;
varying vec2 _gl4es_TexCoord_0;

void main() {
vec4 vertex = _gl4es_ModelViewMatrix * _gl4es_Vertex;
vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);
clippedvertex_0 = dot(vertex, _gl4es_ClipPlane_0);
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=1 Front = 0 Back = 0
Color = _gl4es_FrontMaterial.emission;
Color += _gl4es_Color*_gl4es_LightModel.ambient;
SecColor=vec4(0.);
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_Color.xyz * _gl4es_LightSource_0.ambient.xyz;
nVP = dot(normal, VP);
dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_0.diffuse.xyz):vec3(0.);
hi = normalize(VP + normalize(-vertex.xyz));
lVP = dot(normal, hi);
ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);
Color.rgb += att*(aa+dd);
SecColor.rgb += att*(ss);
// end of light 0
// light 1 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_1.position.xyz);
aa = _gl4es_Color.xyz * _gl4es_LightSource_1.ambient.xyz;
nVP = dot(normal, VP);
dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_1.diffuse.xyz):vec3(0.);
hi = normalize(VP + normalize(-vertex.xyz));
lVP = dot(normal, hi);
ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_1.specular.xyz):vec3(0.);
Color.rgb += att*(aa+dd);
SecColor.rgb += att*(ss);
// end of light 1
Color.a = _gl4es_Color.a;
Color.rgb = clamp(Color.rgb, 0., 1.);
SecColor.rgb = clamp(SecColor.rgb, 0., 1.);
// texturing
// texture 0 active: 1  
_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;
}
```
and Fragment:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000001, textype=00000001, texformat=00000000 point=0
varying vec4 Color;
varying vec4 SecColor;
varying mediump float clippedvertex_0;
varying vec2 _gl4es_TexCoord_0;
uniform sampler2D _gl4es_TexSampler_0;
void main() {
if((min(0., clippedvertex_0))<0.) discard;
vec4 fColor = Color;
vec4 texColor0 = texture2D(_gl4es_TexSampler_0, _gl4es_TexCoord_0);
// Texture 0 active: 1, texenv=0, format=0
fColor *= texColor0;
// Add Secondary color (light )
fColor.rgb += (SecColor).rgb;
fColor.rgb = clamp(fColor.rgb, 0., 1.);
gl_FragColor = fColor;
}
```

Just to note: I assume that what is wrong is the Vertex attribute for the Color (because it seems alpha is null, and alpha is only set by the Color in the shader).
According the capture I have taken, that VA is not update often (because value doesn't change) and Array are disabled, and value is a constant of 1., 1., 0.2, 1. (so yellow with no transparency).

**Comment by kas1e** [comment]
Will check it now .. 

Btw, not that i know topic very much, but as i know the texture2D() function was deprecated years ago, and replaced with texture(). Maybe it worth to replace it in gl4es too ?

**Comment by ptitSeb** [comment]
you mean is the shader? If yes, then yeah, targeting OpenGL 2.x makes most of everything in gl4es deprecated, I know.
I need to , at least, target OpenGL 3.3 Core to be not deprecated, but I'm not there! OpenGL 2.x and deprecated function will be fine for now.

**Comment by kas1e** [comment]
Yeah, i am about generated shaders. I just about that meaning for ogles2.x , and how it handle shaders. Even ogles2.x (if i remember right) made texture2D depricated. But that not so matter anyway , just notice it from top of head :)

**Comment by kas1e** [comment]
A bit late with answer, but we found issue in warp3d, which cause that effect. Its issue in the shader compiler, which in that vertex shader,  falsely assing alpha component as 0 instead of 1. I was in hope it will be fixed fast (so not write there about) , but as for week nothing, i just wrote there to inform you at least . If you doesn't mind, let's that one will be open for a while bit longer, and once we will fix it, we can close it.

Also to other issues we have (in quake/irrlich, etc), i want to setup PPC linux on my amiga, and compile GL4ES for it. So i will need some kind of OpenGLES emulation for linux probably to be able to test gl4es there ..

**Comment by ptitSeb** [comment]
Thanks for the info.
Yeah, we can keep the issue open for traking.

Test on Linux. Yeah why not, You will need to use software MESA I guess. It works (I use it on x86_64 PC to test gl4es), but I hope your machine is powerfull, because it can eat a lot of CPU cycles!

**Comment by kas1e** [comment]
Even if speed will be awfull, it will be not so matter, as my main goal to test it all on big-endian. just not on amigaos, so to exclude or include all the bugs related to architecture. And anyway it will be better to have more platforms for tests

**Comment by kas1e** [comment]
While trying to help our wap3d author to fix that issue, i tryed to build last commit (1572) for amigaos4, and have some compilation errors :)

src/gl/gl.c: In function 'amiga_pre_swap':
src/gl/gl.c:1085: error: too few arguments to function 'blitMainFBO'

src/gl/init.c:6:27: error: ../../version.h: No such file or directory
src/gl/init.c: In function 'initialize_gl4es':
src/gl/init.c:61: error: 'MAJOR' undeclared (first use in this function)
src/gl/init.c:61: error: (Each undeclared identifier is reported only once
src/gl/init.c:61: error: for each function it appears in.)
src/gl/init.c:61: error: 'MINOR' undeclared (first use in this function)
src/gl/init.c:61: error: 'REVISION' undeclared (first use in this function)
src/gl/init.c:262: warning: implicit declaration of function 'snprintf'
src/gl/init.c:262: warning: incompatible implicit declaration of built-in function 'snprintf'
src/gl/init.c:265: warning: incompatible implicit declaration of built-in function 'snprintf'


All other objects compiles fine

**Comment by ptitSeb** [comment]
I just pushed a fix for the 1st problem (the one in `src/gl/gl.c`).

The second issue is not an issue, you just miss the `version.h` file that is on the root of the git repo.

**Comment by kas1e** [comment]
Yep, now all builds and works, thanks.

I add enable for now debug just for fpe_shader.c , and when i run neverball, it generate a looot of shaders of the same type (that can expain that slow-down of the animation for about 3 seconds on running). Just to run a game and then press "exit" from menu right once it appears to be visibly, i count in logs 34 vertex/fragment shaders. A lot .. Through, is it normal that generation of 68 shaders take about 3 seconds on 2ghz cpu ? Or it not depend on the CPU speed, but only on how GPU works ?

I also checking what shaders is generated and strangey can't find in all generated shaders line "Color.a = _gl4es_Color.a;", only "Color.a = gl_Color.a;" or "Color.a = _gl4es_FrontMaterial_alpha;"  Probably generation of that shader was changed a bit on some stage and now it is FrontMaterial_alpha ?

**Comment by ptitSeb** [comment]
All the shader should be different. If you find 2 generated program that are identical (so both vertex and fragment shaders are the same), then there is a bug.

Shader compilation may also depends on the GPU, not only CPU speed.

Now, the shader that you see are "raw" shaders generated by fpe_shader. Those shaders are generated in "OpenGL" style (and should run on actual OpenGL hardware). So it's normal you see stuff like `gl_Color.a`. If you want actual GLES Shaders, you need to see in `src/gl/shaderconv.c` for the converted  version.

**Comment by kas1e** [comment]
Yes all shaders are different indeed. Interesting that today's GPU's are that slow when it come to generation of 64 shaders. I was expected from today's hardware that it all takes milliseconds :)

I also enabed now debug only in shaderconv, and while it didn't look as good as fpe_shader.c debug (where shown what one are vertex, what one are fragment, etc, but in shaderconv i can see only 

:
 (FPEShader generated)
New Shader source:

But at least i can see in all the logs now Color.a = _gl4es_Color.a; line. 

Through question is: how can i found what shader to what place in game is reflect. I.e. i need to find out the smallest possible shader , which is shown that alpha bug, and trying to change some things in.

**Comment by kas1e** [comment]
Btw, and also after just enabling debug in shaderconv, i didn't see comments like you have in shader you post above. I.e. nothing about "// light 0 on, light_direction=0, light_cutoff180=0" for example. 

There is how my debug log looks like when i only enable debug in shaderconv.c:

http://kas1e.mikendezign.com/aos4/gl4es/gles_shaders.txt

**Comment by ptitSeb** [comment]
You need the the environnement variable LIBGL_COMMENTS=1 to have comments in converted shaders...

Seeing what shader are used on what graphic part is a really tricky part. You need to capture the GLES2 command and replay them step by step. On linux there is apitrace (https://apitrace.github.io/ ), not sure it runs on AmigaOS4. On my side, I also use PVRTrace. I still have a PVRTrace capture of neverball, so if you tell me what you are looking for, I can try to help.

**Comment by kas1e** [comment]
If you have time, plz check this: http://amigadeveloper.com/bugreports/view.php?id=318

At end of bz Hans ask to replace some things in that shader to see if bug disappear, and to have small test case for. So as test case seems not that easy to write (at least fast), i thinking about finding that shader and replace that alpha value he ask for begining. Also was in hope to find less complecated shader which will show same issue (that why i ask about to which parts shaders refers)..

Or maybe you will have anything to add to BZ so Hans can find roots faster

Thanks for your time, again :)

**Comment by ptitSeb** [comment]
I'll take a look (but I have to create an account for this first, can't browse the ticket as anonymous).

**Comment by ptitSeb** [comment]
Ok, I see the bug repport.

If you want to be sure, in `src/gl/fpe_shader.c` line 469:
replace
`sprintf(buff, "Color.a = %s;\n", (color_material && (state->cm_front_mode==FPE_CM_DIFFUSE || state->cm_front_mode==FPE_CM_AMBIENTDIFFUSE))?"gl_Color.a":"_gl4es_FrontMaterial_alpha");`
with
`sprintf(buff, "Color = %s;\n", (color_material && (state->cm_front_mode==FPE_CM_DIFFUSE || state->cm_front_mode==FPE_CM_AMBIENTDIFFUSE))?"gl_Color":"vec4(_gl4es_FrontMaterial_alpha)");`
To force rgba assignment instead of just "a".
Once you prove this work, change again with:
`sprintf(buff, "Color = %s;\n", (color_material && (state->cm_front_mode==FPE_CM_DIFFUSE || state->cm_front_mode==FPE_CM_AMBIENTDIFFUSE))?"vec4(gl_Color.a)":"vec4(_gl4es_FrontMaterial_alpha)");`

To see if problem is indeed with the reading of `gl_Color.a` (that is realy `_gl4es_Color` and indeed a VA) or the assignment to `Color.a`.

**Comment by ptitSeb** [comment]
Mmmm, I have redone a PVRTrace capture of latest neverball, using latest gl4es.

For the record, here is what the 1st level looks like on the Pandora:
![neverball_ok](https://user-images.githubusercontent.com/3250792/39956912-ec898d78-55e9-11e8-9e8f-adb442a705b5.png)

(notice the yellow halo thingy at the finish point).

Now, the coins are drawn using Material, but no color material (contrary to what I wrote earlier). The shaders are:
vertex:
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
uniform highp mat3 _gl4es_NormalMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute highp vec4 _gl4es_MultiTexCoord0;
attribute highp vec3 _gl4es_Normal;
struct _gl4es_LightModelProducts
{
   vec4 sceneColor;
};
uniform _gl4es_LightModelProducts _gl4es_FrontLightModelProduct;
uniform _gl4es_LightModelProducts _gl4es_BackLightModelProduct;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=1 (twosided=0, separate=1, color_material=0)
// secondary=0, planes=000000
// texture=00000001, point=0
varying vec4 Color;
struct _gl4es_FPELightSourceParameters1
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
   highp float constantAttenuation;
   highp float linearAttenuation;
   highp float quadraticAttenuation;
};
struct _gl4es_FPELightSourceParameters0
{
   highp vec4 specular;
   highp vec4 position;
   highp vec3 spotDirection;
   highp float spotExponent;
   highp float spotCosCutoff;
};
struct _gl4es_LightProducts
{
   highp vec4 ambient;
   highp vec4 diffuse;
   highp vec4 specular;
};
uniform highp float _gl4es_FrontMaterial_shininess;
uniform highp float _gl4es_FrontMaterial_alpha;
uniform highp float _gl4es_BackMaterial_alpha;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;
uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;
uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;
varying vec4 SecColor;
varying vec2 _gl4es_TexCoord_0;

void main() {
vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
// ColorMaterial On/Off=0 Front = 0 Back = 0
Color = _gl4es_FrontLightModelProduct.sceneColor;
SecColor=vec4(0.);
highp float att;
highp float spot;
highp vec3 VP;
highp float lVP;
highp float nVP;
highp vec3 aa,dd,ss;
highp vec3 hi;
// light 0 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_0.position.xyz);
aa = _gl4es_FrontLightProduct_0.ambient.xyz;
nVP = dot(normal, VP);
dd = (nVP>0.)?(nVP * _gl4es_FrontLightProduct_0.diffuse.xyz):vec3(0.);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);
Color.rgb += att*(aa+dd);
SecColor.rgb += att*(ss);
// end of light 0
// light 1 on, light_direction=0, light_cutoff180=0
att = 1.0;
VP = normalize(_gl4es_LightSource_1.position.xyz);
aa = _gl4es_FrontLightProduct_1.ambient.xyz;
nVP = dot(normal, VP);
dd = (nVP>0.)?(nVP * _gl4es_FrontLightProduct_1.diffuse.xyz):vec3(0.);
hi = normalize(VP + vec3(0., 0., 1.));
lVP = dot(normal, hi);
ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial_shininess)*_gl4es_FrontLightProduct_1.specular.xyz):vec3(0.);
Color.rgb += att*(aa+dd);
SecColor.rgb += att*(ss);
// end of light 1
Color.a = _gl4es_FrontMaterial_alpha;
Color.rgb = clamp(Color.rgb, 0., 1.);
SecColor.rgb = clamp(SecColor.rgb, 0., 1.);
// texturing
vec4 tmp_tex;
// texture 0 active: 1  
_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy / _gl4es_MultiTexCoord0.q;
}
```
and fragment:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=1, alpha=0, secondary=0, planes=000000, textype=00000001, texformat=00000001 point=0
varying vec4 Color;
varying vec4 SecColor;
varying vec2 _gl4es_TexCoord_0;
uniform sampler2D _gl4es_TexSampler_0;
void main() {
vec4 fColor = Color;
vec4 texColor0 = texture2D(_gl4es_TexSampler_0, _gl4es_TexCoord_0);
// Texture 0 active: 1, texenv=0, format=1
fColor.rgb *= texColor0.rgb;
// Add Secondary color (light )
fColor.rgb += (SecColor).rgb;
fColor.rgb = clamp(fColor.rgb, 0., 1.);
gl_FragColor = fColor;
}
```
The changes I put above will probably not test correctly.. For hans's test, line 469 should be:
`sprintf(buff, "Color.a = %s;\n", (color_material && (state->cm_front_mode==FPE_CM_DIFFUSE || state->cm_front_mode==FPE_CM_AMBIENTDIFFUSE))?"gl_Color.a":"1.0");`

**Comment by kas1e** [comment]
Thanks, will check now. I do all tests on neverball 1.5.4, so it may be a little different than your version, but probably in terms of shaders will be the same.

**Comment by ptitSeb** [comment]
Ah yeah, it was 1.5.4. I'll redo the capture on this version (I have this one too) and see if the rendering is the same or not.

**Comment by ptitSeb** [comment]
And yes, 1.5.4 use Color Material, and the latest version from github doesn't. So the 1st shader is the correct one for your tests.
(I note that the rendering of latest github version is more efficient then the 1.5.4, less rendering command are generated with latest version then with the 1.5.4)

**Comment by kas1e** [comment]
Tested, with changed 469 line on what you wrote in last post, so i have in shaders Color.a = gl_Color.a; and that didn't change things.

But (!) if i change that line instead on pure sprintf(buff, "Color.a = 1.0;\n"); , so i have in shaders Color.a = 1.0; , then it works, and coins displays fine. Go figure :)

**Comment by kas1e** [comment]
In intro i have it too, also almost transparent, but visibly. Also, when i wait enough in the intro, first level starts as demo, and it shows the yellow thingy and final point. But if i start first for real, then it for sure completely ignored and show up only when i collect enough coins to exit from level. 

Watched video as well , yeah .. seems something with alha on warp3d side seems so..

**Comment by kas1e** [comment]
With changing whole color values all works. Coins here.

Daniel in his report says that :

// This here doesn't work. Color.a will end up being 0.

Color.a = _gl4es_Color.a;

// However _gl4es_Color contains an a != 0.
// If you use the following line here instead the one above, Color.a will be okay:

Color = _gl4es_Color;

**Comment by ptitSeb** [comment]
Strange.

I can write a simple temporary workaround for this bug in gl4es (basically the code you have now, but with `#ifdef AMIGAOS`), that will allow you to test more things while waiting for a fix.

Now that yollow glow missing is probably another issue.
The shader use for this one is dead simple:
vertex:
```
#version 100
precision mediump float;
precision mediump int;
uniform highp mat4 _gl4es_ModelViewProjectionMatrix;
attribute highp vec4 _gl4es_Vertex;
attribute lowp vec4 _gl4es_Color;
// FPE_Shader generated
// ** Vertex Shader **
// ligthting=0 (twosided=0, separate=0, color_material=0)
// secondary=0, planes=000000
// texture=00000000, point=0
varying vec4 Color;

void main() {
gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;
Color = _gl4es_Color;
// texturing
}
```
and fragment:
```
#version 100
precision mediump float;
precision mediump int;
// FPE_Shader generated
// ** Fragment Shader **
// lighting=0, alpha=0, secondary=0, planes=000000, textype=00000000, texformat=00000000 point=0
varying vec4 Color;
void main() {
vec4 fColor = Color;
gl_FragColor = fColor;
}
```
and the Attributes used are VA of 4*GL_FLOAT. It should work flawlessly, that's strange you don't have it.

**Comment by kas1e** [comment]
I also downloaded 1.5.4 windows version (to check how yellow thingy works in the finish point), and there is also no yellow thingy in first level until i didn't collect enough coins. So probably that expected, and maybe in last versions they do change something ?

**Comment by ptitSeb** [comment]
Ok, my bad.
In fact, the yellow glow + stars are there at start only if you alread have beaten the level !

Sorry for the false alarm.

**Comment by kas1e** [comment]
As for workoround, nah, that not need , as it cleary bug in warp3d driver and should be fixed on our side. I just want somehow to help Hans to find out roots :) And it anyway happens to be only in neverball at moment.

But what i also notice in that 1.5.4 windows version, is that coins are mirrored in the intro:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverball_1_5_4_mirrored_coins.jpg

Did you have it on pandora ?

**Comment by ptitSeb** [comment]
Yes. There are mirrored also on the Pandora screenshot I have put here.

**Comment by ptitSeb** [comment]
I have pushed the workaround. It's temporary, only affect AMIGAOS4 build, and will allow you to test other stuff.
I'll remove it as soon as it's not needed anymore.
You still don't have coins reflection in the intro?

**Comment by kas1e** [comment]
Right, indeed. For me it didn't even when i use Color.a = 1.0; and have working coins.

I also noticed that the "bridges" in the intro (those parts which connect those big letters) in windows version are transparent, same as for you. As well, as whole letters also a bit transparent, and i can see in windows version starts over them.

For me, those "bridges" are solid. As well as big letters not transparent. Check this out:
http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/non_transparent_bridges.jpg

**Comment by ptitSeb** [comment]
I assume that's because so forced Color.a = 1.0
Try with latest source (so using my workaround), to see if you have coins, reflections, and transparent bridges.

**Comment by ptitSeb** [comment]
I don't think this is the issue here.
The color is a float, so it's a 32bits value. There is no non-32bits value involved here. In fact, I don't think gl4es will send any non-32bits value to GLES2 driver anyway. The only case that could happens is the use of GL_UNSIGNED_BYTE and only 3 values for colors, but gl4es will transform that to 4-value FLOAT anyway, as actuel GLES2 drivers don't handle that well anyway.

There is something else going on...

**Comment by kas1e** [comment]
Doesn't you mind if i will send that information to bugreport too ?

**Comment by kas1e** [comment]
What you think, if take that shader in question, and use it with some simple triangle drawing, can it show issue for us somehow ? Reducing test case for sure will help him.

**Comment by ptitSeb** [comment]
Not a problem.
Also, I have done a post on the bug traker. That GPU assembly stuff is tasty :)

**Comment by ptitSeb** [comment]
Yes you can. But it will require some setup, the shader is pretty complex and use a lot of uniform.

**Comment by kas1e** [comment]
Lets see what Hans will say and when , before starting with test cases. That for sure will be easy for him to find roots once some simple test case in use.

Maybe it even possible to make shorten version of the buggy shader before (i.e. cut off parts till it will show the bug), and once it will short enough, use it with test case

**Comment by ptitSeb** [comment]
Ok. In the mean time, you have a fully working version of neveball (and neverputt I guess), using GLES2 and all. How the speed (graphic quality also) compare to minigl?

**Comment by kas1e** [comment]
Speed is MUCH faster. Which kind of surprise when we cheat that much with quake3 to be at least on the same level.  For example first level of neverball give me stable 60-62fps without problems for g4es version , While minigl one give me about 30fps.

In neverputt differences is much bigger: First level of neverputt give me 60fps in the minigl version, while give me 160fps for gl4es version.

When i wrote this before to Daniel (ogles2 author), he says that maybe neverXXXX uses display-lists which probably results in a less inefficient render-path inside gl4es, or maybe there is VBOs uses.  

But then, we tried Q3 before with VBOs and have no differences, so something else involved which make neverball be much faster.

**Comment by kas1e** [comment]
Found few issues with neverputt. Check plz if it the same or not on pandora.

Problems is two. First one: border of table didn't have drawned triangles, instead it like solid color too. And second one, is that "hole" looks like solid color too. On windows version, and on minigl version all is ok in that terms. There is 3 screens.

win32: http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_hole_win32.jpg

amigaos4/minigl: http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_hole_minigl.jpg

amigaos4/gl4es: http://kas1e.mikendezign.com/aos4/gl4es/games/neverball/neverputt_hole_gl4es.jpg

**Comment by ptitSeb** [comment]
I think a speed difference can be explained with glTexGen. There are used in places with neverball/neverputt, and they executed in shaders with gl4es where the are certainly emulated in software with minigl.

**Comment by ptitSeb** [comment]
Neverputt looks fine on my side
![neverputt_ok](https://user-images.githubusercontent.com/3250792/39958374-48fcf3ac-5602-11e8-9872-e613620b95aa.png)
I guess this is another manifestation of ticket 318.
I'll do a PVRTrace of this to see how this drawn and what are the shaders used.

**Comment by ptitSeb** [comment]
The rendering is slightly different then with neverball indeed.
Quick question: did you had the white hole / very bright light before the workaround?

**Comment by ptitSeb** [comment]
Nevermind, I have the anwser: no (there is a screenshot earlier).
Do you have any shader compilation error?

**Comment by ptitSeb** [comment]
I have pushed a modified version of the workaround.
I still do `Color = gl_Color;` before light computation in case of color material, but in the other case (like here with neverputt), it keep the older `Color.a = _gl4es_FrontMaterial_alpha;`

**Comment by kas1e** [comment]
Yep, latest change for workoround make neverlputt works as before, and make neverball be all-right as well. Now let's see what Hans answer :) We bring him some more new info, so maybe something new will come from it.

**Comment by ptitSeb** [comment]
Well, that make it looks much more complicated than that is initialy look. Maybe that's the shader compiler that sometimes unexpectdly overwrite a value. We'll see indeed.

**Comment by kas1e** [comment]
Yep..

Btw Hans ask if we can disable completely lighting in shader to see if bug still here, so it will be much smaller test case. But for me problem is that how we can know what one of those 34 generated shaders we can use to reproduce a bug ?:)

From another side we can disable lighting at all (for all shaders), then to see if bug still here, and then, without lighting trying to create small test case (if it will be easy than with lighting) ?

**Comment by ptitSeb** [comment]
Well, as a quick hack, you can change the vertex shader generated by simply changing it line 652 of `src/gl/fpe_shader.c` and line 1123 for the fragment shader.
You have some sort of signature in the shaders code I copied on my early post https://github.com/ptitSeb/gl4es/issues/57#issuecomment-376649353
For example insert that:
```
    if(lighting==1 && alpha==0 && secondary==0 && planes==1 && textype==1 && texformat==0 && point==0) {
        strcpy(shad, 
            "#version 100\n"
            "precision mediump float;\n"
            "precision mediump int;\n"
            "// FPE_Shader generated\n"
            "// ** Fragment Shader **\n"
            "// lighting=1, alpha=0, secondary=0, planes=000001, textype=00000001, texformat=00000000 point=0\n"
            "varying vec4 Color;\n"
            "varying vec4 SecColor;\n"
            "varying mediump float clippedvertex_0;\n"
            "varying vec2 _gl4es_TexCoord_0;\n"
            "uniform sampler2D _gl4es_TexSampler_0;\n"
            "void main() {\n"
            "if((min(0., clippedvertex_0))<0.) discard;\n"
            "vec4 fColor = Color;\n"
            "vec4 texColor0 = texture2D(_gl4es_TexSampler_0, _gl4es_TexCoord_0);\n"
            "// Texture 0 active: 1, texenv=0, format=0\n"
            "fColor *= texColor0;\n"
            "// Add Secondary color (light )\n"
            "fColor.rgb += (SecColor).rgb;\n"
            "fColor.rgb = clamp(fColor.rgb, 0., 1.);\n"
            "gl_FragColor = fColor;\n"
            "}\n"        
        );
    }
```
(that the same shader, but that way, you can still change stuff)

But more important, the overide the vertex shader with (insert in line 652)
```
    if(lighting==1 && twosided==0 && separate==1 && color_material==1 && secondary==0 && planes==1 && texture==1 && point==0 && light==3) {
        "#version 100\n"
        "precision mediump float;\n"
        "precision mediump int;\n"
        "uniform highp mat4 _gl4es_ModelViewMatrix;\n"
        "uniform highp mat4 _gl4es_ModelViewProjectionMatrix;\n"
        "uniform highp mat3 _gl4es_NormalMatrix;\n"
        "attribute highp vec4 _gl4es_Vertex;\n"
        "attribute lowp vec4 _gl4es_Color;\n"
        "attribute highp vec4 _gl4es_MultiTexCoord0;\n"
        "attribute highp vec3 _gl4es_Normal;\n"
        "struct _gl4es_LightModelParameters {\n"
        "  vec4 ambient;\n"
        "};\n"
        "uniform _gl4es_LightModelParameters _gl4es_LightModel;\n"
        "struct _gl4es_MaterialParameters\n"
        "{\n"
        "   vec4 emission;\n"
        "   vec4 ambient;\n"
        "   vec4 diffuse;\n"
        "   vec4 specular;\n"
        "   float shininess;\n"
        "};\n"
        "uniform _gl4es_MaterialParameters _gl4es_FrontMaterial;\n"
        "uniform _gl4es_MaterialParameters _gl4es_BackMaterial;\n"
        "// FPE_Shader generated\n"
        "// ** Vertex Shader **\n"
        "// ligthting=1 (twosided=0, separate=1, color_material=1)\n"
        "// secondary=0, planes=000001\n"
        "// texture=00000001, point=0\n"
        "varying vec4 Color;\n"
        "uniform highp vec4 _gl4es_ClipPlane_0;\n"
        "varying mediump float clippedvertex_0;\n"
        "struct _gl4es_FPELightSourceParameters1\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "   highp float constantAttenuation;\n"
        "   highp float linearAttenuation;\n"
        "   highp float quadraticAttenuation;\n"
        "};\n"
        "struct _gl4es_FPELightSourceParameters0\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "};\n"
        "struct _gl4es_LightProducts\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "};\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;\n"
        "varying vec4 SecColor;\n"
        "varying vec2 _gl4es_TexCoord_0;\n"
        "\n"
        "void main() {\n"
        "vec4 vertex = _gl4es_ModelViewMatrix * _gl4es_Vertex;\n"
        "vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);\n"
        "clippedvertex_0 = dot(vertex, _gl4es_ClipPlane_0);\n"
        "gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;\n"
        "// ColorMaterial On/Off=1 Front = 0 Back = 0\n"
        "Color = _gl4es_FrontMaterial.emission;\n"
        "Color += _gl4es_Color*_gl4es_LightModel.ambient;\n"
        "SecColor=vec4(0.);\n"
        "highp float att;\n"
        "highp float spot;\n"
        "highp vec3 VP;\n"
        "highp float lVP;\n"
        "highp float nVP;\n"
        "highp vec3 aa,dd,ss;\n"
        "highp vec3 hi;\n"
        "// light 0 on, light_direction=0, light_cutoff180=0\n"
        "att = 1.0;\n"
        "VP = normalize(_gl4es_LightSource_0.position.xyz);\n"
        "aa = _gl4es_Color.xyz * _gl4es_LightSource_0.ambient.xyz;\n"
        "nVP = dot(normal, VP);\n"
        "dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_0.diffuse.xyz):vec3(0.);\n"
        "hi = normalize(VP + normalize(-vertex.xyz));\n"
        "lVP = dot(normal, hi);\n"
        "ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);\n"
        "Color.rgb += att*(aa+dd);\n"
        "SecColor.rgb += att*(ss);\n"
        "// end of light 0\n"
        "// light 1 on, light_direction=0, light_cutoff180=0\n"
        "att = 1.0;\n"
        "VP = normalize(_gl4es_LightSource_1.position.xyz);\n"
        "aa = _gl4es_Color.xyz * _gl4es_LightSource_1.ambient.xyz;\n"
        "nVP = dot(normal, VP);\n"
        "dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_1.diffuse.xyz):vec3(0.);\n"
        "hi = normalize(VP + normalize(-vertex.xyz));\n"
        "lVP = dot(normal, hi);\n"
        "ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_1.specular.xyz):vec3(0.);\n"
        "Color.rgb += att*(aa+dd);\n"
        "SecColor.rgb += att*(ss);\n"
        "// end of light 1\n"
        "Color.a = _gl4es_Color.a;\n"
        "Color.rgb = clamp(Color.rgb, 0., 1.);\n"
        "SecColor.rgb = clamp(SecColor.rgb, 0., 1.);\n"
        "// texturing\n"
        "// texture 0 active: 1  \n"
        "_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;\n"
        "}\n"
    }
```

Those are unmodified shaders. Just try it like that to confirm you have the bug back, and that everything else work as intented. Then you can simplify the vertex shader...

**Comment by kas1e** [comment]
Something seems wrong with syntax there, i insert them both and have:

src/gl/fpe_shader.c: In function 'fpe_VertexShader':
src/gl/fpe_shader.c:653: error: 'separate' undeclared (first use in this function)
src/gl/fpe_shader.c:653: error: (Each undeclared identifier is reported only once
src/gl/fpe_shader.c:653: error: for each function it appears in.)
src/gl/fpe_shader.c:653: error: 'texture' undeclared (first use in this function)
src/gl/fpe_shader.c:653: error: 'light' undeclared (first use in this function)
src/gl/fpe_shader.c:769: error: expected ';' before '}' token
src/gl/fpe_shader.c: In function 'fpe_FragmentShader':
src/gl/fpe_shader.c:1244: error: 'alpha' undeclared (first use in this function)
src/gl/fpe_shader.c:1244: error: 'textype' undeclared (first use in this function)
src/gl/fpe_shader.c:1244: error: 'texformat' undeclared (first use in this function)

**Comment by kas1e** [comment]
And probably in vertex one part missed strcpy(shad, as in fragment one ?

**Comment by ptitSeb** [comment]
Yeah, sorry, I typed a bit too fast and forgot things...

here is vertex shader part:
```
    if(lighting==1 && twosided==0 && light_separate==1 && color_material==1 && secondary==0 && planes==1 && state->textype==1 && point==0 && state->light==3) {
        strcpy(shad,
        "#version 100\n"
        "precision mediump float;\n"
        "precision mediump int;\n"
        "uniform highp mat4 _gl4es_ModelViewMatrix;\n"
        "uniform highp mat4 _gl4es_ModelViewProjectionMatrix;\n"
        "uniform highp mat3 _gl4es_NormalMatrix;\n"
        "attribute highp vec4 _gl4es_Vertex;\n"
        "attribute lowp vec4 _gl4es_Color;\n"
        "attribute highp vec4 _gl4es_MultiTexCoord0;\n"
        "attribute highp vec3 _gl4es_Normal;\n"
        "struct _gl4es_LightModelParameters {\n"
        "  vec4 ambient;\n"
        "};\n"
        "uniform _gl4es_LightModelParameters _gl4es_LightModel;\n"
        "struct _gl4es_MaterialParameters\n"
        "{\n"
        "   vec4 emission;\n"
        "   vec4 ambient;\n"
        "   vec4 diffuse;\n"
        "   vec4 specular;\n"
        "   float shininess;\n"
        "};\n"
        "uniform _gl4es_MaterialParameters _gl4es_FrontMaterial;\n"
        "uniform _gl4es_MaterialParameters _gl4es_BackMaterial;\n"
        "// FPE_Shader generated\n"
        "// ** Vertex Shader **\n"
        "// ligthting=1 (twosided=0, separate=1, color_material=1)\n"
        "// secondary=0, planes=000001\n"
        "// texture=00000001, point=0\n"
        "varying vec4 Color;\n"
        "uniform highp vec4 _gl4es_ClipPlane_0;\n"
        "varying mediump float clippedvertex_0;\n"
        "struct _gl4es_FPELightSourceParameters1\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "   highp float constantAttenuation;\n"
        "   highp float linearAttenuation;\n"
        "   highp float quadraticAttenuation;\n"
        "};\n"
        "struct _gl4es_FPELightSourceParameters0\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "};\n"
        "struct _gl4es_LightProducts\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "};\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;\n"
        "varying vec4 SecColor;\n"
        "varying vec2 _gl4es_TexCoord_0;\n"
        "\n"
        "void main() {\n"
        "vec4 vertex = _gl4es_ModelViewMatrix * _gl4es_Vertex;\n"
        "vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);\n"
        "clippedvertex_0 = dot(vertex, _gl4es_ClipPlane_0);\n"
        "gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;\n"
        "// ColorMaterial On/Off=1 Front = 0 Back = 0\n"
        "Color = _gl4es_FrontMaterial.emission;\n"
        "Color += _gl4es_Color*_gl4es_LightModel.ambient;\n"
        "SecColor=vec4(0.);\n"
        "highp float att;\n"
        "highp float spot;\n"
        "highp vec3 VP;\n"
        "highp float lVP;\n"
        "highp float nVP;\n"
        "highp vec3 aa,dd,ss;\n"
        "highp vec3 hi;\n"
        "// light 0 on, light_direction=0, light_cutoff180=0\n"
        "att = 1.0;\n"
        "VP = normalize(_gl4es_LightSource_0.position.xyz);\n"
        "aa = _gl4es_Color.xyz * _gl4es_LightSource_0.ambient.xyz;\n"
        "nVP = dot(normal, VP);\n"
        "dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_0.diffuse.xyz):vec3(0.);\n"
        "hi = normalize(VP + normalize(-vertex.xyz));\n"
        "lVP = dot(normal, hi);\n"
        "ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_0.specular.xyz):vec3(0.);\n"
        "Color.rgb += att*(aa+dd);\n"
        "SecColor.rgb += att*(ss);\n"
        "// end of light 0\n"
        "// light 1 on, light_direction=0, light_cutoff180=0\n"
        "att = 1.0;\n"
        "VP = normalize(_gl4es_LightSource_1.position.xyz);\n"
        "aa = _gl4es_Color.xyz * _gl4es_LightSource_1.ambient.xyz;\n"
        "nVP = dot(normal, VP);\n"
        "dd = (nVP>0.)?(nVP * _gl4es_Color.xyz * _gl4es_LightSource_1.diffuse.xyz):vec3(0.);\n"
        "hi = normalize(VP + normalize(-vertex.xyz));\n"
        "lVP = dot(normal, hi);\n"
        "ss = (nVP>0. && lVP>0.)?(pow(lVP, _gl4es_FrontMaterial.shininess)*_gl4es_FrontLightProduct_1.specular.xyz):vec3(0.);\n"
        "Color.rgb += att*(aa+dd);\n"
        "SecColor.rgb += att*(ss);\n"
        "// end of light 1\n"
        "Color.a = _gl4es_Color.a;\n"
        "Color.rgb = clamp(Color.rgb, 0., 1.);\n"
        "SecColor.rgb = clamp(SecColor.rgb, 0., 1.);\n"
        "// texturing\n"
        "// texture 0 active: 1  \n"
        "_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;\n"
        "}\n"
        );
    }
```

and here the fragment shader part
```
    if(lighting==1 && alpha_test==0 && secondary==0 && planes==1 && state->textype==1 && state->texformat==0 && point==0) {
        strcpy(shad, 
            "#version 100\n"
            "precision mediump float;\n"
            "precision mediump int;\n"
            "// FPE_Shader generated\n"
            "// ** Fragment Shader **\n"
            "// lighting=1, alpha=0, secondary=0, planes=000001, textype=00000001, texformat=00000000 point=0\n"
            "varying vec4 Color;\n"
            "varying vec4 SecColor;\n"
            "varying mediump float clippedvertex_0;\n"
            "varying vec2 _gl4es_TexCoord_0;\n"
            "uniform sampler2D _gl4es_TexSampler_0;\n"
            "void main() {\n"
            "if((min(0., clippedvertex_0))<0.) discard;\n"
            "vec4 fColor = Color;\n"
            "vec4 texColor0 = texture2D(_gl4es_TexSampler_0, _gl4es_TexCoord_0);\n"
            "// Texture 0 active: 1, texenv=0, format=0\n"
            "fColor *= texColor0;\n"
            "// Add Secondary color (light )\n"
            "fColor.rgb += (SecColor).rgb;\n"
            "fColor.rgb = clamp(fColor.rgb, 0., 1.);\n"
            "gl_FragColor = fColor;\n"
            "}\n"        
        );
    }
```

**Comment by kas1e** [comment]
Thanks ! Compiles fine, but on running of neverball crashes , with words:
FPE Vertex hader compie failed: ERROR: 96:2 '' : syntax error, unexpected $end
ERROR: 1 compilation errors. No code generated.

**Comment by ptitSeb** [comment]
Ok, that was a bit to rough...

In both block, change
`strcpy(shad, `
with
```
        strcpy(shad, "");
        ShadAppend(
```
That will fix buffer overfow issues.

**Comment by kas1e** [comment]
Yep, that works. Neverball starts, all seems works as before, except that there is no "mirrored" coins in the intro (but coins itself here, and bridges also transparent, etc, just no mirrors of coins). Should't i remove those amigaos ifdefs , or , they can stay ?

If i understand right, for now we return always only one shader of the same type ?

**Comment by ptitSeb** [comment]
No the hack doesn't change all shaders, but only 1: the one used for the coins.
In fact, it changes more then 1 shaders, that's why it was crashing before.

Now, the mirrored coins are the one worth looking at: it should be the graphics object this forced shader change (because this one shader trigger the W3D, so alpha will be 0).
You can now start experiementing things. Like put the test hans asked, by changing the vertex shader too
```
        "#version 100\n"
        "precision mediump float;\n"
        "precision mediump int;\n"
        "uniform highp mat4 _gl4es_ModelViewMatrix;\n"
        "uniform highp mat4 _gl4es_ModelViewProjectionMatrix;\n"
        "uniform highp mat3 _gl4es_NormalMatrix;\n"
        "attribute highp vec4 _gl4es_Vertex;\n"
        "attribute lowp vec4 _gl4es_Color;\n"
        "attribute highp vec4 _gl4es_MultiTexCoord0;\n"
        "attribute highp vec3 _gl4es_Normal;\n"
        "struct _gl4es_LightModelParameters {\n"
        "  vec4 ambient;\n"
        "};\n"
        "uniform _gl4es_LightModelParameters _gl4es_LightModel;\n"
        "struct _gl4es_MaterialParameters\n"
        "{\n"
        "   vec4 emission;\n"
        "   vec4 ambient;\n"
        "   vec4 diffuse;\n"
        "   vec4 specular;\n"
        "   float shininess;\n"
        "};\n"
        "uniform _gl4es_MaterialParameters _gl4es_FrontMaterial;\n"
        "uniform _gl4es_MaterialParameters _gl4es_BackMaterial;\n"
        "// FPE_Shader generated\n"
        "// ** Vertex Shader **\n"
        "// ligthting=1 (twosided=0, separate=1, color_material=1)\n"
        "// secondary=0, planes=000001\n"
        "// texture=00000001, point=0\n"
        "varying vec4 Color;\n"
        "uniform highp vec4 _gl4es_ClipPlane_0;\n"
        "varying mediump float clippedvertex_0;\n"
        "struct _gl4es_FPELightSourceParameters1\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "   highp float constantAttenuation;\n"
        "   highp float linearAttenuation;\n"
        "   highp float quadraticAttenuation;\n"
        "};\n"
        "struct _gl4es_FPELightSourceParameters0\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "   highp vec4 position;\n"
        "   highp vec3 spotDirection;\n"
        "   highp float spotExponent;\n"
        "   highp float spotCosCutoff;\n"
        "};\n"
        "struct _gl4es_LightProducts\n"
        "{\n"
        "   highp vec4 ambient;\n"
        "   highp vec4 diffuse;\n"
        "   highp vec4 specular;\n"
        "};\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_0;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_0;\n"
        "uniform _gl4es_FPELightSourceParameters0 _gl4es_LightSource_1;\n"
        "uniform _gl4es_LightProducts _gl4es_FrontLightProduct_1;\n"
        "varying vec4 SecColor;\n"
        "varying vec2 _gl4es_TexCoord_0;\n"
        "\n"
        "void main() {\n"
        "vec4 vertex = _gl4es_ModelViewMatrix * _gl4es_Vertex;\n"
        "vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);\n"
        "clippedvertex_0 = dot(vertex, _gl4es_ClipPlane_0);\n"
        "gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;\n"
        "Color.a = _gl4es_Color.a;\n"
        "Color.rgb = _gl4es_Color.rgb;\n"
        "SecColor.rgb = vec3(0.);\n"
        "// texturing\n"
        "// texture 0 active: 1  \n"
        "_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;\n"
        "}"
```

**Comment by kas1e** [comment]
That one fail to compiles:

VectorShuffle instruction with a constant parameter isn't supported yet.
INTERNA ERROR: Store instruction's source parameter is incompete:
 OpStore: tmp59>>SecColor

**Comment by kas1e** [comment]
But neverball starts anyway, and show me the same working intro with all the stuff working, just without mirrored coins.

**Comment by ptitSeb** [comment]
Mmm, then change
`"SecColor.rgb = vec3(0.);\n"`
with
`"SecColor = vec4(0.);\n"`

**Comment by kas1e** [comment]
With that change no errors from shader, but, i don't see mirrored coins in intro still. 

For sake of test i comment out our shaders insertion (just to see if all works without) : yes, without i can see mirrored coins.

**Comment by ptitSeb** [comment]
Ok
and if change
```
        "Color.a = _gl4es_Color.a;\n"
        "Color.rgb = _gl4es_Color.rgb;\n"
```
with
```
        "Color =vec4(1.)\n"
```

Do you see white reflected coins?

**Comment by kas1e** [comment]
Yes, with that change i see them.

**Comment by ptitSeb** [comment]
Ok, now changed that
```
        "Color =vec4(1.)\n"
```
with
```
        "Color = _gl4es_Color;\n"
```
Yellow reflected coins?

**Comment by kas1e** [comment]
Yes, now mirrored coins are yellow

**Comment by ptitSeb** [comment]
last test,
after
```
        "Color = _gl4es_Color;\n"
```
add back
```
        "Color.a = _gl4es_Color.a;\n"
```
Reflected coins are gone?

**Comment by kas1e** [comment]
Yep, they gone now :)

**Comment by kas1e** [comment]
Yes, Yellow reflected coins are back

**Comment by ptitSeb** [comment]
Ok.
Mybe one more last test then, to be sure.
Before the end of the shader, add back
```
        "Color.a = _gl4es_Color.a;\n"
```
To see if the reflected coins are gone again or still there.

**Comment by kas1e** [comment]
Added it at end, right after "_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;\n", and mirrored coins are gone again.

**Comment by ptitSeb** [comment]
Ok. I think hans will have enything he needs now.
It seems the `Color.a = _gl4es_Color.a;` somehow doesn't work, while `Color = _gl4es_Color;` does work, or `Color.a = 1.0;` works too. And doesn't seem to be dependent on what is after.

The simplest maybe to copy/paster each vertex shader with the result to the hans's ticket I guess.

**Comment by kas1e** [comment]
Yeah, that what Daniel find before and what he report originally, and that seems not enough for hans :(

Or you mean now shaders without lighting , so it shoud be easer for him ?

**Comment by kas1e** [comment]
Btw, probably i can remove from our shader all those struct and uniforms too ? (as they used for lighting, which we didn't use in main() anymore ?)

**Comment by kas1e** [comment]
Remove all non used parts from shader, and that what i have now (bug can be triggered still):

		"#version 100\n"
        "precision mediump float;\n"
        "precision mediump int;\n"
        "uniform highp mat4 _gl4es_ModelViewMatrix;\n"
        "uniform highp mat4 _gl4es_ModelViewProjectionMatrix;\n"
        "uniform highp mat3 _gl4es_NormalMatrix;\n"
        "attribute highp vec4 _gl4es_Vertex;\n"
        "attribute lowp vec4 _gl4es_Color;\n"
        "attribute highp vec4 _gl4es_MultiTexCoord0;\n"
        "attribute highp vec3 _gl4es_Normal;\n"
		
        "// FPE_Shader generated\n"
        "// ** Vertex Shader **\n"
        "// ligthting=1 (twosided=0, separate=1, color_material=1)\n"
        "// secondary=0, planes=000001\n"
        "// texture=00000001, point=0\n"
        "varying vec4 Color;\n"
        "uniform highp vec4 _gl4es_ClipPlane_0;\n"
        "varying mediump float clippedvertex_0;\n"
        "varying vec4 SecColor;\n"
        "varying vec2 _gl4es_TexCoord_0;\n"
        "\n"
        "void main() {\n"
        "vec4 vertex = _gl4es_ModelViewMatrix * _gl4es_Vertex;\n"
        "vec3 normal = normalize(_gl4es_NormalMatrix * _gl4es_Normal);\n"
        "clippedvertex_0 = dot(vertex, _gl4es_ClipPlane_0);\n"
        "gl_Position = _gl4es_ModelViewProjectionMatrix * _gl4es_Vertex;\n"
		
		// don't work !
		//"Color.a = _gl4es_Color.a;\n"
		
		// works :
		//"Color.a = 1.0;\n"
		
		// works too : 
		//"Color = _gl4es_Color;\n"
				
		"SecColor = vec4(0.);\n"
        "// texturing\n"
        "// texture 0 active: 1  \n"
        "_gl4es_TexCoord_0 = _gl4es_MultiTexCoord0.xy;\n"		
        "}"


Probably can just copy+paste that

**Comment by kas1e** [comment]
With such a small shader maybe it possible to made a test case for , like drawing triangle with some alpha or so ?

**Comment by ptitSeb** [comment]
Yep.

Don't forget the `Color.rgb=_gl4es_Color.rgb;` that can be before the `Color.a=_gl4es_Color.a` and also the test to do the affectation as the last instruction.

**Comment by kas1e** [comment]
Even without Color.rgb = _gl4es_Color.rgb bug can be enabled/disabled. Or that need to be noted as well ?

**Comment by ptitSeb** [comment]
Yes, because this assignment `Color.rgb = _gl4es_Color.rgb` works (the coins are yellow when using  `Color.a=1.0;`)

**Comment by kas1e** [comment]
Ok, i made 2 posts to BZ , take a look plz if i describe everything well enough and there is no needs to add anything else. And thanks for help with our crap :)

**Comment by ptitSeb** [comment]
Yep, I have seen them. Looks good to me. Let's wait hans answer and analysis now.

**Comment by kas1e** [comment]
@ptitSeb

Btw, is that error in shader when we use in shader SecColor.rgb = vec3(0.);  instead of SecColor = vec4(0.); 
, which produce that " VectorShuffle instruction with a constant parameter isn't supported yet. INTERNAL ERROR: Store instruction's source parameter is incompete: OpStore: tmp59>>SecColor" , mean our shader compiler need to be updated, or, it was just some general shader error and not depent on amigaos4 only ? (i just want to create another feature request for it on our BZ, if it the case).

**Comment by ptitSeb** [comment]
I think it needs to be updated. Setting .rgb with a constant is legal.
What I'm not sure is if the error is because of the `SecColor.rgb = vec3(0.);` or because `SecColor.a` is not assigned anything? Try with
```
SecColor.rgb = vec3(0.);
SecColor.a = 0.;
```
To be sure.

**Comment by kas1e** [comment]
Yeah, same error, even with that 2 lines. So probably can just copy error text to feature request.

**Comment by ptitSeb** [comment]
Yep.

**Comment by ptitSeb** [comment]
@kas1e : It seems ticket 318 is fixed. If you can confirm it is, I'll remove the workaround and we can close this ticket.

**Comment by kas1e** [comment]
Yeah, once will get new warp3d and confirm that all fine will write you back there (i didn't get new versions of warpd right away, so it can take few days more).

**Comment by kas1e** [comment]
Tested new warp3d : yeah, it is fixed. We can close ticket and remove workaround from gl4es. Thanks !

**Comment by ptitSeb** [comment]
Workaround removed with 5b3a077c500d5f5f80986bae825623f6239f4505

**Comment by ptitSeb** [comment]
So Neverball & neverputt are the 1st games that works perfectly and are faster than with MiniGL \o/

Will you package it for AmigaOS, so other can try it (and maybe find new bugs on specifics configuration) ?

**Comment by kas1e** [comment]
Yeah, they for sure faster.  And yeah, i will made release archive as well soon. 

At moment i thinking how to enable extensions to amigaos4 version in right way. I.e. we before add it, but in console we firstly have "LIBGL: Hardware test disabld, nothing activated..." , and after some time extensions enables, which somehow reads unlogical. Maybe we need just ifndef that part for amigaos4 about printf ? And then, add it to the place when it call in amigaos4, so all will reads logical in console ?

**Comment by kas1e** [comment]
Thanks ! Can you also add another cosmetic : enable those "glXetProcAddress not founded.." stubs for amigaos4 too , remember i have them all when run aos4 version of irrlicht's helloworld test case:
http://kas1e.mikendezign.com/aos4/gl4es/irrlicht/irrlicht_helloworld.jpg
They do nothing as you say, but still will be also good to have it same "silent" as in other versions of gl4es.

About neverball/neverputt release : i can't release it now, as those new fixed warp3d/ogles2 , is not publicaly released yet (i as betatester have it), so i need to wait for new public release first :(  In meantime will try to recude quake3 sources maybe, to be able to short test case for that "normalization / ubyte" bug.

**Comment by ptitSeb** [comment]
The LIBGL_SILENTSTUB=1 is not working for you or you want that to be default on AmigaOS?

**Comment by kas1e** [comment]
Oh yes, LIBGL_SILENTSTUB=1 works. I made false assumption befoer that it default for all other platforms except amigaos4, but if it the same everywhere then all is fine, thanks for worry :)

**Comment by ptitSeb** [comment]
Yes, it's ON on all plateform. It's a usefull indicator on what (in term of OpenGL functionnality) the software is trying to use.

**Comment by kas1e** [comment]
Btw, in new version of our warp3d was fixed some issues in shaders, so, i am able to compile and run neverball 1.6.0 now, yeah ! 

And what i notice now, that by default there is no menu, and speed of first animation scene are 40fps.

Then, when scene runs second time, all drops to 15 fps. Like something going wrong after it shows first time.

Then after some experiments i found, that settings LIBGL_BATCH to 1 , make menu be visibly (but not that much, only letters), and speed in intro scene 25 fps (as letters has to be drawn), but at least when animation go second time, there is not fps descreasing.

Then, LIBGL_BATCH 2 looks the same as LIBGL_BATCH 1 , and by speed and by not full look of menu.

Then, LIBGL_BATCH 3, give full looks of menu , FPS are 47, and didn't drops when animation starts second time. 

Then i tried LIBGL_BATCH 20, same as 3, (and 4 and 5, all same).

Then LIBGL_BATCH 50,  fps drops to 40-42 fps (so start to be slower than LIBGL_BATCH 20).

And when do LIBGL_BATCH 100, its the same as 50.


So, the best value probabaly 20 .

Question is : is it ok ? I mean is it the same on Pandora that by default you see no menu and have issue with animation played second time, and only with LIBGL_BATCH 20 , etc sort it all ? Should't it works by default anyway ?

Also game seems leak on exit graphics memory (i hope it game), but that can be probabaly as usuall something like missing destroying of SDL context and all that proper exit in the game's code, need to check.

**Comment by kas1e** [comment]
Ah, and another question : if it all like this, i then need for neverball to setup LIBGL_BATCH 20 in the gl4es code specially, how to do it easy way ? I mean making special static gl4es library with LIBGL_BATCH 20 as default, so neverball will works fine right from the start

**Comment by ptitSeb** [comment]
Mmmm, I don' have this at all.
I don't need any `LIBGL_BATCH` to fix the rendering. So there is something wrong here. Can you try with `LIBGL_BEGINEND=0` just to check if it's an issue with the Merger?
I need to retry on the Pandora with 1.60 specifically, but last I try I had no issue at all.

**Comment by kas1e** [comment]
Tried with LIBGL_BEGINEND=0 : no differences. Same. Only when LIBGL_BATCH is set to more than 3 all start to be correct. I use very latest gl4es (well, maybe minus commits about new stub functions).

Is merger works when LIBGL_BATCH is used ? Just tryied to set LIBGL_BATCH to 20, and then disable merger, and it make no differences visually, the same as with enabled merger.

Btw, leak on exit was because they just use SDL_Quit without proper cleaning of context/window before, added that and leak gone.

I have pretty much impressive FPS numbers in compare with old 1.5.4 version and with minigl/gl4es in whole, check this out , window mode, 1024x768 , some first level with doing nothing, for gl4es LIBGL_BATCH 20:

```
Neverball:

v1.5.4 minigl: 28 FPS
v1.5.4 gl4es : 55 FPS
v1.6.2 minigl : can't build, not enough GL function support
v1.6.2 gl4es : 82 FPS
```
```
Neverputt:

```
v1.5.4 minigl: 60 FPS
v1.5.4 gl4es : 300 FPS
v1.6.2 minigl : can't build, not enough GL function support
v1.6.2 gl4es : 310 FPS
```

Not bad :)

**Comment by ptitSeb** [comment]
Well `LIBGL_BATCH` and the Merger act on different part (glDrawArrars/glDrawElements and glBegin/glEnd respectively), but there maybe some interection between the two. I need to recheck the code of Batch mode.
The FPS are indeed impressive!

**Comment by ptitSeb** [comment]
I quickly tested latest gl4es with neverball 1.6 on an Linux VM (using a simple apt get neverball...), and I have menu and all on first run, no issue here.

**Comment by kas1e** [comment]
Interesting.. I use one from their site downloaded az 1.6.0 source code, tgen datafiles from win32 version on top of it.

Is for you when you set libgl_batch 20 things start to be much faster than by default ?

Interesting what can be different there that i have no menu when no batching used..  I mean is it bug or feature :)

I probably anyway need setup libgl_batch 20 hardcore in gl4es for neverball, as it speed things a lot as well (just where better to do it?)

**Comment by ptitSeb** [comment]
I haven't printed the fps, so I cannot say if LIBGL_BATCH=20 speeds things up. It can spped up, that's the only purpose of LIBGL_BATCH in fact. 
But that fact that you need some LIBGL_BATCH is not ok. There is something wrong happening without LIBGL_BATCH (but yeah, I cannot reproduce it om my side).

Also, there is a github repo of neverball (https://github.com/Neverball/neverball ). Why not using this?

**Comment by kas1e** [comment]
I just didn't use githib repo as very latest code can be usually with problems/bugs , while stable releases kind of stable (at leat in theory :) ).

I may try also github's version, but as i see from log , they do almost nothing important since 1.6.0.

Through strange, what can cause that , and more of it how to catch what wrong with ..

**Comment by kas1e** [comment]
Btw, is setting of LIBGL_BATCH 20 need to be done inside of init.c , just instead of globals4es.batch = 0;  put globals4es.batch = 20; ?  It just not that i mean it for fix , but as it for sure speed things up, so i want it be default for neverball/neverputt builds

**Comment by ptitSeb** [comment]
Yes, to activate default `LIBGL_BATCH=20` you need to do that.

To catch what's wrong, yes, that will be tricky. This seems somehow still related to "small" rendering. But that also can just be a coinsidence of course. In fact, I have no clue...

**Comment by kas1e** [comment]
It can be something about internal optimisation of ogles2 / warp3dnova.. Or it can be just some side effect of memory trashing bug somewhere (and in neverball itself for example, just, as usuall, on amigaos4 every single bug is visibly and can make a harm)..

**Comment by ptitSeb** [comment]
Well, because it can be anything, it's difficult to track. And I don't think internal optimisation in ogles2 / warp3dnova can be disabled, so, it's even more difficult to say.
I don't remember neverball beeing too buggy (were foobillard++ have some bugs in the engine for sure!).

**Comment by kas1e** [comment]
It can be that i use win32 data files ..Will just try to build very latest one, with all datafiles, etc, so if it will be the same, then let it be like this, at least libgl_batch 20 will fix everything and it cool to have it working that fast

**Comment by kas1e** [comment]
With latest source the same story, so probabaly be it. Maybe if it bug somewhere it will showups soon or later somewhere else where will be more easy to detect it. At least it works and for release will be good enough :)

**Comment by ptitSeb** [comment]
Sure, I understand, no problem! Enjoy Neverball @60fps+ and neverputt @300fps :O !
